### PR TITLE
add a new json dataformat for fastjson library

### PR DESCRIPTION
**Pull request by kimmking**
add a new json dataformat for fastjson library:
https://github.com/alibaba/fastjson

### CODE REVIEW

**Comment by davsclaus** [comment]
Thanks, do you mind fix checkstyle issues

    mvn compile -P sourcecheck

File: components/camel-fastjson/src/main/java/org/apache/camel/component/fastjson/FastjsonDataFormat.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+                    config.getCharset(), //
+                    graph, //
+                    config.getSerializeConfig(), //
+                    config.getSerializeFilters(), //
+                    config.getDateFormat(), //
+                    JSON.DEFAULT_GENERATE_FEATURE, //
+                    config.getSerializerFeatures());
+
+        if (contentTypeHeader) {
+            if (exchange.hasOut()) {
```
Comment (onders86) [inline components/camel-fastjson/src/main/java/org/apache/camel/component/fastjson/FastjsonDataFormat.java]: More NPE safe as below i would say

Message message = null;
if (exchange.hasOut()) {
	message = exchange.getOut();
} else {
	message = exchange.getIn();
}
if (message != null) {
	message.setHeader(Exchange.CONTENT_TYPE, "application/json");
	message.setHeader(Exchange.CONTENT_LENGTH,len);
}

**Comment by kimmking** [comment]
fixed now.
thx.
@davsclaus @onders86

File: components/camel-fastjson/src/main/java/org/apache/camel/component/fastjson/FastjsonDataFormat.java [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        int len = JSON.writeJSONString(stream,
+                config.getCharset(),
+                graph,
+                config.getSerializeConfig(),
+                config.getSerializeFilters(),
+                config.getDateFormat(),
+                JSON.DEFAULT_GENERATE_FEATURE,
+                config.getSerializerFeatures());
+
+        if (contentTypeHeader) {
```
Comment (davsclaus) [inline components/camel-fastjson/src/main/java/org/apache/camel/component/fastjson/FastjsonDataFormat.java]: Just a note that exchange is never null, also that getIn or getOut also is never null (they lazy create a new empty message)

File: components/camel-fastjson/src/main/java/org/apache/camel/component/fastjson/FastjsonDataFormat.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                serializerFeatureList.add(SerializerFeature.PrettyFormat);
+            }
+            if (serializeNulls) {
+                serializerFeatureList.add(SerializerFeature.WriteMapNullValue);
+                serializerFeatureList.add(SerializerFeature.WriteNullBooleanAsFalse);
+                serializerFeatureList.add(SerializerFeature.WriteNullListAsEmpty);
+                serializerFeatureList.add(SerializerFeature.WriteNullNumberAsZero);
+                serializerFeatureList.add(SerializerFeature.WriteNullStringAsEmpty);
+            }
+            if (null != this.dateFormatPattern) {
```
Comment (davsclaus) [inline components/camel-fastjson/src/main/java/org/apache/camel/component/fastjson/FastjsonDataFormat.java]: We tend to flip this so != null is last

**Comment by davsclaus** [comment]
This looks good anyway +1

**Comment by onders86** [comment]
New data format API is cool contribution. thanks. One last thing may be asking you to squash your commits?

**Comment by oscerd** [comment]
I'm a bit concerned about the license first line:

https://github.com/alibaba/fastjson/blob/master/license.txt

**Comment by kimmking** [comment]
Licensed under the Apache License, Version 2.0 (the "License");

@oscerd

**Comment by oscerd** [comment]
I was looking at the copyright

Copyright 1999-2016 Alibaba Group Holding Ltd.

I guess it's not a problem, but maybe @davsclaus can answer better than me.

**Comment by kimmking** [comment]
@davsclaus  i fixed it

**Comment by kimmking** [comment]
@davsclaus  does this copyright in ref library matter? thx.

@onders86  how to squash commits? can you give me some tips, thx.

File: components/camel-fastjson/pom.xml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    <dependencies>
+
+        <dependency>
+            <groupId>org.apache.camel</groupId>
+            <artifactId>camel-core</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.alibaba</groupId>
+            <artifactId>fastjson</artifactId>
+            <version>1.2.32</version>
```
Comment (dmvolod) [inline components/camel-fastjson/pom.xml]: Please set version as variable inside camel-parent/pom.xml

Comment (dmvolod) [inline components/camel-fastjson/pom.xml]: Please set version as variable inside camel-parent/pom.xml

**Comment by davsclaus** [comment]
No the (c) is not a problem, its fine. There are many ASL2 projects that are licensed by a company, for example Hazelcast is also ASL2 license but (c) from that company
https://github.com/hazelcast/hazelcast/blob/master/hazelcast-client/src/main/java/com/hazelcast/client/ClientExtension.java#L2

**Comment by davsclaus** [comment]
Its okay about squashing, sadly we don't have a nice way of doing this from the github website, which you can do in some projects. We can merge this as-is.
As if you have not do any squash before it can be a bit danger to do, if you don't do it correct. I have a alias for my git client so I can squash X commits back. I found it somewhere on the internet
```
squash	 => !f(){ git reset --soft HEAD~${1} && git commit --edit -m"$(git log --format=%B --reverse HEAD..HEAD@{1})"; };f
```


@kimmking next contribution, then remember to create a JIRA ticket and refer to the CAMEL-NNN number of the JIRA in the commit messages.

File: components/camel-fastjson/pom.xml (hunk 4 above)
Comment (kimmking) [inline components/camel-fastjson/pom.xml]: fixed it, thx
@dmvolod

**Comment by kimmking** [comment]
@davsclaus  thx so much. I have squashed commits now.

And thanks a lot to every nice guy here.

**Comment by wuwen5** [comment]
@kimmking 
https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request#squash-your-changes

**Comment by kimmking** [comment]
@wuwen5  thx too much.

**Comment by davsclaus** [comment]
Thanks for the contribution

**Pull request closed without merging**